contracts/transmuter/src/test/cases/units/join_and_exit.rs

```diff
@@ -0,0 +1,557 @@
+use std::collections::HashMap;
+
+use cosmwasm_std::{attr, Coin, Uint128};
+use osmosis_test_tube::{Account, OsmosisTestApp};
+
+use crate::{
+    contract::{
+        ExecMsg, GetShareDenomResponse, GetSharesResponse, GetTotalPoolLiquidityResponse,
+        GetTotalSharesResponse, InstantiateMsg, QueryMsg,
+    },
+    test::test_env::{assert_contract_err, TestEnvBuilder},
+    ContractError,
+};
+
+#[test]
+fn test_join_pool_with_single_lp_should_update_shares_and_liquidity_properly() {
+    #[derive(Debug)]
+    struct Case {
```

Comment:
love the setup on this test. It's so clean. Much better than the test-case crate, imo